website/blog/unusual-builds-w-bytes.md

```diff
@@ -0,0 +1,217 @@
+
+As we digged deeper into the issue, we tried to improve the visibility of Prefetch actions in Bazel with [PR#25040](https://github.com/bazelbuild/bazel/pull/25040), which was backported to Bazel 8.1.0 and 7.5.0.
+This change breaks down the `prefetcher` action ID into "input" downloads and "output" downloads, allowing us to narrow down the code paths that were triggering the unwanted downloads.
+
+Once we had [upgraded to a Bazel version](./blog/bisect-bazel) with this change, we were able to see that the unwanted downloads were coming from the `output` downloads.
+
+# The Bug
+
+We discovered another code path in Bazel that can trigger unexpected downloads: when a blob's TTL expires.
+
+> So... What is a blob's TTL? 🤔
+
+As an artifact-oriented build system, Bazel relies heavily on artifacts existing in its output path to determine if a build is up-to-date.
+If an artifact exists and matches the expected result, then Bazel will skip the action that produces the artifact.
+However, in a remote build with minimal download, Bazel does not have the artifact locally to validate against and must establish some trust that the artifact was created and stored in the Remote Cache.
+
+Introduced in [PR#17639](https://github.com/bazelbuild/bazel/pull/17639), this "trust" is added to Bazel by storing the expected Time To Live (TTL) of the artifact in the Remote Cache.
+When Bazel executes an action remotely, instead of downloading the outputs referenced inside the ActionResult, it instead just stores the references inside its in-process analysis cache (aka. SkyFrame) with an expected TTL that is determined by the flag `--experimental_remote_cache_ttl` (default: 3 hours).
+When the TTL expires, instead of checking if the output artifacts are up-to-date remotely, Bazel downloads the entire artifact to disk.
+We were able to validate this by setting the flag `--experimental_remote_cache_ttl=0s` to eagerly trigger the downloads.
+
+```bash
+# Initial build
+$ bazel clean && bazel test //cli/... --config=remote-minimal --disk_cache= --experimental_remote_cache_ttl=1d
+$ du -h $(bazel info output_base)/execroot | sort -h
+...
+1.8M	/private/var/tmp/_bazel_fmeum/412888b82b4f18156bc415025cb8faa1/execroot
+
+# Incremental build
+$ bazel clean && bazel test //cli/... --config=remote-minimal --disk_cache= --experimental_remote_cache_ttl=1d
+$ du -h $(bazel info output_base)/execroot | sort -h
+...
+1.8M	/private/var/tmp/_bazel_fmeum/412888b82b4f18156bc415025cb8faa1/execroot
+
+# Incremental build with TTL expired
+$ bazel clean && bazel test //cli/... --config=remote-minimal --disk_cache= --experimental_remote_cache_ttl=0
+$ du -h $(bazel info output_base)/execroot | sort -h
+...
+640M	/private/var/tmp/_bazel_fmeum/412888b82b4f18156bc415025cb8faa1/execroot
+```
+
+There was a draft fix by [David Sanderson](https://github.com/dws) in [PR#23066](https://github.com/bazelbuild/bazel/pull/23066), however, it was incomplete and never reviewed.
```

Comment:
Interesting, I wasn't aware of this PR. It does work around the issue, but its ultimate aim is something else.